WIP: grpc interface

hammadzz opened the pull request:
<!--- Make sure PR title follows contentional-commits pattern -->

## Description
Here is an early pilot (see To Do) for gRPC interface that gives a parallel gateway to accessing Wemmick.

Currently gRPC covers the following APIs:
1. HelloWorld (perhaps we can switch this to a simple health check)
1. ListDataSources
1. ListExpectationSuites
1. CreateExpectationSuiteFromJsonSchema

## To Do:
- [x] figure out pip dependency management (perhaps remove pipenv)
- [x] test gRPC with docker
- [x] remove or mask credentials in connection_string from ListDataSources response
- [ ] refactor protobuf namespace and follow better practices on naming of service
- [x] decide if a better name is suited for api.py and refactor to share code with cli
- [x] add remaining CLI features to gRPC interface

## Closes issue(s)
[Kanbanize 751](https://bera.kanbanize.com/ctrl_board/9/cards/751/details/)

## How to test / repro
N/A

## Screenshots
N/A

## Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

## Checklist

- [x] I have tested this code
- [x] I have updated the Readme

## Other comments
None

Aylr commented on src/wemmick/grpc_server.py:
```diff
@@ -0,0 +1,58 @@
+                name=data_source_item['name'],
+                class_name=data_source_item['class_name'],
+                module_name=data_source_item['module_name'],
+                credentials=credentials
```
We probably don't want to publish credentials.

hammadzz replied:
Yup I remember the discomfort as I wrote that line. I added a todo in the description as wanted to decide if we drop this altogether or use some sort or regex to mask the secrets but not the host and username.

The credentials definition is also incomplete. It currently only expects `connection_url`.

tomdavidson replied:
I think wemmick can check env and creds file:
 
ENV=prod|dev|ect
CREDENTIALS_FILE_PATH=creds.yaml

As for consumers of wemmick, they can vol mount or build a new image.  I want to use sops for secret change management via git. kan-428

BUT this is an open source project, and we dont put any db config/cred in it.

hammadzz replied:
This api returns the datasource's connection_url (nested in credentials). I was hoping to parse the connection_url and mask the secret while returning the db name and table name.  I removed the credentials field for now. Will have to leave that for later.

Aylr commented on src/wemmick/api.py:
```diff
@@ -0,0 +1,34 @@
+import json
```
I love that you built an `api` file. Nice

Aylr commented on src/wemmick/protos/wemmick.proto:
```diff
@@ -0,0 +1,42 @@
+}
+
+message DataSource {
+  message Credentials {
+    string connection_string = 1;
+  };
```
```suggestion
```
No need to include this.

hammadzz replied:
Has been removed!

Aylr commented on src/wemmick/grpc_server.py:
```diff
@@ -0,0 +1,70 @@
+
+        return wemmick_pb2.ExpectationSuitesList(expectation_suites=result)
+
+    def CreateExpectationSuiteFromJsonSchema(self, request, context):
+        create_suite = wemmick.api.CreateExpectationSuiteFromJsonSchema(
+            file_path=request.file_path,
+            suite_name=request.suite_name
+        )
+        create_suite.run()
+        return wemmick_pb2.Response(message=f'Successfully created {request.suite_name} expectation suite')
+
+    def CreateExpectationSuiteFromAvroSchema(self, request, context):
+        create_suite = wemmick.api.CreateExpectationSuiteFromAvroSchema(
+            file_path=request.file_path,
+            suite_name=request.suite_name
+        )
+        create_suite.run()
+        return wemmick_pb2.Response(message=f'Successfully created {request.suite_name} expectation suite')
```
I'm likely missing something about the intended use case here, and/or have not communicated the intent of wemmick (which is dual purpose - a server that can be triggered to run a validation and an interactive CLI tool to create suites):

My impression was that the grpc interface was for triggering validations, not for creation of suites which would require passing files around. From a UX perspective it's going to be way easier to use the wemmick CLI to create suites than to invoke wemmick via grpc and shuttles files around. I'd motion that we remove all but the validation functionality.

hammadzz replied:
My vision has been along the lines of Wemmick being a service hosted in your cloud provider which anyone in the team could create new suites from schemas, run validations with these suites, and view results. All without ever having to install even python on their local machine. So my scope for [751](https://bera.kanbanize.com/ctrl_board/9/cards/751/details/) has been to give cli equivalent functionality in gRPC

This allows to programatically create the latest expectation suites from a schema profiler before running a validation operator. So you would just have to point to any schema file and be able to validate a table against it.

Aylr replied:
That makes sense to me. Thanks for the clarification.

tomdavidson replied:
I think your both right on :D 

But the MVP the gRPC interface only needs to invoke existing suites and JSON Schema generated ones.

We can expand what the interface does in additional iterations which is why I want gRPC. I want the commands and types to be "discovered" by the CLI.


Also, the REST interface just needs to do the same invoking but via from Cloud Scheduler.

hammadzz commented:
@Aylr currently this repo does not have any GE project. I assume a dummy GE project will be needed for unit testing?

Currently a unit test is failing. This unit test checks to ensure an error is thrown on missing json file. As my change calls `ge.DataContext()` call before check for missing json file it breaks the test. I could reorder missing json file check to fix it but I expect we will need a dummy GE project for other unit tests.

Aylr commented:
Good call out on the tests. I think the best solution here would be to either mock the call to data context or make an in-memory fixture rather than include an entire dummy project just for this test. I'm happy to help with that if you'd like.

tomdavidson commented:
If there is a mock or fixture, make sure it is actually testing the interface and not GE end to end style - and that there is not a way do it without mocks or fixtures, even if that requires code restructure.

hammadzz commented:
Ready to review again. No longer a draft.

Updates:
- Added a few basic unit tests for common api and one for gRPC, this required a lot of refactoring on common api
- Ensured common api unit tests are on the interface only, GE api calls are mocked / faked out
- Setup grpc server to run in Docker (had some trouble here, python file needed shebang to run)
- Refactored protos directory (had some trouble here, this is the only way to get module imports instead of relative imports on generated stubs)

Will leave the additional unit testing for gRPC for future improvements. Haven't found good examples on how to test just the gRPC interface.

tomdavidson commented on src/wemmick/utils.py:
```diff
@@ -14,3 +15,36 @@ def file_relative_path(dunderfile, relative_path: str):
+        # TODO: add GCP Bucket support
+        pass
+    elif scheme == "file":
+        file_path = urlparse(path).path
```
I think it would be better if this logic was its own function outside of the elif

hammadzz replied:
Will refactor this

hammadzz replied:
Cleaned this up.

tomdavidson commented on src/wemmick/utils.py:
```diff
@@ -14,3 +15,36 @@ def file_relative_path(dunderfile, relative_path: str):
+def get_file(path: str):
+    scheme = get_scheme(path)
+    if scheme == "s3":
+        # TODO: add AWS S3 support
```
id rather not have todos like this in the code ...  say we had a ci action that created kanban cards for all todos (which would be quite nice) - we wouldnt want to create that card for maybes.

hammadzz replied:
I will remove these

tomdavidson commented on src/wemmick/utils.py:
```diff
@@ -14,3 +15,36 @@ def file_relative_path(dunderfile, relative_path: str):
+
+def get_file(path: str):
+    scheme = get_scheme(path)
+    if scheme == "s3":
```
Can python do a map and reducer like in js instead of the elif ? (just currious)

hammadzz replied:
Will look more into this

hammadzz replied:
So far I found [this pattern](https://hackernoon.com/how-to-properly-use-switch-case-statements-in-python-explained-with-examples-ivg536vx). Will consider it when adding S3 and GCS options.

tomdavidson replied:
Nice, that was a great reference and helped my understanding python a bit more.  I particularly like the dict mapping to functions.

hammadzz commented on src/wemmick/cli.py:
```diff
@@ -99,47 +89,78 @@ def avro_glob(pattern: str, verbose: bool = False):
     typer.secho(f"Processed {len(suites)} avro files.", fg=typer.colors.BRIGHT_GREEN)
 
 
+# TODO: add missing functionality that was present in cli code that was refactored
```
I compared it now. I added most of the functionality, one thing missing is the verbose option.

hammadzz replied:
Added verbose option back and removed todo comment.
